### PR TITLE
Gateway: Paginate SearchRoutes API

### PR DESCRIPTION
**Pull request by dbczumar**
## What changes are proposed in this pull request?

Gateway: Paginate SearchRoutes API

## How is this patch tested?

- [ ] Existing unit/integration tests
- [X] New unit/integration tests - **NEEDS TESTS**
- [ ] Manual tests (describe details, including test results, below)

## Does this PR change the documentation?

- [X] No. You can skip the rest of this section.
- [ ] Yes. Make sure the changed pages / sections render correctly in the documentation preview.

### Is this a user-facing change?

- [X] No. You can skip the rest of this section.
- [ ] Yes. Give a description of this change to be included in the release notes for MLflow users.

(Details in 1-2 sentences. You can just refer to another PR with a description if this PR is part of a larger change.)

### What component(s), interfaces, languages, and integrations does this PR affect?

Components

- [ ] `area/artifacts`: Artifact stores and artifact logging
- [ ] `area/build`: Build and test infrastructure for MLflow
- [ ] `area/docs`: MLflow documentation pages
- [ ] `area/examples`: Example code
- [ ] `area/model-registry`: Model Registry service, APIs, and the fluent client calls for Model Registry
- [X] `area/models`: MLmodel format, model serialization/deserialization, flavors
- [ ] `area/recipes`: Recipes, Recipe APIs, Recipe configs, Recipe Templates
- [ ] `area/projects`: MLproject format, project running backends
- [ ] `area/scoring`: MLflow Model server, model deployment tools, Spark UDFs
- [ ] `area/server-infra`: MLflow Tracking server backend
- [ ] `area/tracking`: Tracking Service, tracking client APIs, autologging

Interface

- [ ] `area/uiux`: Front-end, user experience, plotting, JavaScript, JavaScript dev server
- [ ] `area/docker`: Docker use across MLflow's components, such as MLflow Projects and MLflow Models
- [ ] `area/sqlalchemy`: Use of SQLAlchemy in the Tracking Service or Model Registry
- [ ] `area/windows`: Windows support

Language

- [ ] `language/r`: R APIs and clients
- [ ] `language/java`: Java APIs and clients
- [ ] `language/new`: Proposals for new client languages

Integrations

- [ ] `integrations/azure`: Azure and Azure ML integrations
- [ ] `integrations/sagemaker`: SageMaker integrations
- [ ] `integrations/databricks`: Databricks integrations

<a name="release-note-category"></a>

### How should the PR be classified in the release notes? Choose one:

- [ ] `rn/breaking-change` - The PR will be mentioned in the "Breaking Changes" section
- [X] `rn/none` - No description will be included. The PR will be mentioned only by the PR number in the "Small Bugfixes and Documentation Updates" section
- [ ] `rn/feature` - A new user-facing feature worth mentioning in the release notes
- [ ] `rn/bug-fix` - A user-facing bug fix worth mentioning in the release notes
- [ ] `rn/documentation` - A user-facing documentation change worth mentioning in the release notes

### CODE REVIEW

File: mlflow/gateway/client.py [hunk 1]
```diff
@@ -95,26 +96,25 @@ def get_route(self, name: str):
...
 
-        .. note::
-            Search is currently not implemented. Providing a `search_filter` term will raise an
-            exception. Leave the arguments empty to get a listing of all configured routes.
 
         :param search_filter: An optional filter to apply to the search. Currently not used.
         :return: Returns a list of all configured and initialized `Route` data for the MLflow
             Gateway Server. The return will be a list of dictionaries that detail the name, type,
             and model details of each active route endpoint.
         """
```
Comment (BenWilson2) [inline mlflow/gateway/client.py:108]: silly nit: pydoc entry for page_token + remove entry for search_filter

**Review by BenWilson2 — APPROVED** [review]
Looks solid! :)

File: mlflow/gateway/client.py (hunk 1 above)
Comment (dbczumar) [reply mlflow/gateway/client.py:108]: Thanks for the catch - updated docstrings :)

File: mlflow/gateway/config.py [hunk 2]
```diff
@@ -240,12 +240,24 @@ def to_route(self) -> "Route":
         )
 
 
-class Route(BaseModel, extra=Extra.allow):
```
Comment (dbczumar) [inline mlflow/gateway/config.py:243]: `extra=Extra.allow` is incompatible with `Config`. Looking into https://docs.pydantic.dev/1.10/usage/model_config/#extra-attributes more, I think we should be using `extra=Extra.ignore` anyway, else attributes from updated backends will start to surface in old clients, potentially causing confusion

Comment (dbczumar) [reply mlflow/gateway/config.py:243]: cc @BenWilson2 @harupy I'll file another PR to change to `Extra.ignore`

File: mlflow/gateway/providers/anthropic.py [hunk 3]
```diff
@@ -49,7 +49,8 @@ async def completions(self, payload: completions.RequestPayload) -> completions.
         if candidate_count != 1:
             raise HTTPException(
                 status_code=422,
-                detail="'candidate_count' must be '1' for the Anthropic provider, if specified.",
+                detail="'candidate_count' must be '1' for the Anthropic provider. "
+                f"Received value: '{candidate_count}'.",
```
Comment (dbczumar) [inline mlflow/gateway/providers/anthropic.py:53]: Follow-up from https://github.com/mlflow/mlflow/pull/8938#discussion_r1251294054. Accidentally clicked merge before addressing

File: mlflow/gateway/config.py (hunk 2 above)
Comment (dbczumar) [reply mlflow/gateway/config.py:243]: Filed https://github.com/mlflow/mlflow/pull/8940 to address

File: mlflow/gateway/app.py [hunk 4]
```diff
@@ -137,10 +173,20 @@ async def get_route(route_name: str) -> Route:
         )
 
     @app.get(MLFLOW_GATEWAY_CRUD_ROUTE_BASE)
-    async def search_routes() -> SearchRoutesResponse:
-        # placeholder route listing functionality
-
-        return {"routes": list(app.dynamic_routes.values())}
+    async def search_routes(page_token: Union[str, None] = None) -> SearchRoutesResponse:
```
Comment (harupy) [inline mlflow/gateway/app.py:176]:
```suggestion
    async def search_routes(page_token: Optional[str] = None) -> SearchRoutesResponse:
```

for consistency

File: tests/gateway/test_client.py [hunk 5]
```diff
@@ -229,6 +230,43 @@ def test_search_mixed_routes(mixed_gateway):
...
+            :MLFLOW_GATEWAY_NUM_ROUTES_PER_SEARCH_RESULTS_PAGE
+        ]
+        assert routes_page_1.token
+
+        routes_page_2 = gateway_client.search_routes(page_token=routes_page_1.token)
+        assert len(routes_page_2) == 5
+        assert [route.name for route in routes_page_2] == gateway_route_names[
+            MLFLOW_GATEWAY_NUM_ROUTES_PER_SEARCH_RESULTS_PAGE:
+        ]
+        assert not routes_page_2.token
```
Comment (harupy) [inline tests/gateway/test_client.py:267]:
```suggestion
        assert routes_page_2.token is None
```

`not routes_page_2.token` makes me wonder what the expected value is.

File: mlflow/gateway/app.py [hunk 6]
```diff
@@ -137,10 +173,20 @@ async def get_route(route_name: str) -> Route:
...
+            start_idx = 0
+
+        end_idx = start_idx + MLFLOW_GATEWAY_NUM_ROUTES_PER_SEARCH_RESULTS_PAGE
+        routes = list(app.dynamic_routes.values())
+        result = {"routes": routes[start_idx:end_idx]}
+        if len(routes[end_idx:]) > 0:
+            next_page_token = SearchRoutesToken(index=end_idx)
+            result["next_page_token"] = next_page_token.encode()
+
+        return result
```
Comment (harupy) [inline mlflow/gateway/app.py:189]: /ai what is the fastest way to extract the first N values from a dictionary in Python?

File: mlflow/gateway/constants.py [hunk 7]
```diff
@@ -2,6 +2,7 @@
 MLFLOW_GATEWAY_CRUD_ROUTE_BASE = "/api/2.0/gateway/routes/"
 MLFLOW_GATEWAY_ROUTE_BASE = "/gateway/"
 MLFLOW_QUERY_SUFFIX = "/invocations"
+MLFLOW_GATEWAY_NUM_ROUTES_PER_SEARCH_RESULTS_PAGE = 3000
```
Comment (harupy) [inline mlflow/gateway/constants.py:5]: /ai The following constant is used by `mlflow.search_routes` (which supports pagiation) and represents the page size. Suggest 5 names for this constant.

File: mlflow/gateway/app.py (hunk 6 above)
Comment (harupy) [reply mlflow/gateway/app.py:189]: @mlflow thanks!
@dbczumar Never mind this comment

File: mlflow/gateway/constants.py (hunk 7 above)
Comment (harupy) [reply mlflow/gateway/constants.py:5]: I'd use `MLFLOW_GATEWAY_SEARCH_ROUTES_PAGE_SIZE`.

File: mlflow/gateway/utils.py [hunk 8]
```diff
@@ -94,3 +96,39 @@ def assemble_uri_path(paths: List[str]) -> str:
...
+        except Exception as e:
+            raise MlflowException.invalid_parameter_value(
+                f"Invalid SearchRoutes token: {encoded_token}"
+            ) from e
+
+        index = parsed_token.get("index")
+        if not index or index < 0:
+            raise MlflowException.invalid_parameter_value(
+                f"Invalid SearchRoutes token: {encoded_token}"
+            ) from e
```
Comment (harupy) [inline mlflow/gateway/utils.py:123]: Is `0` an invalid `index` value?

File: mlflow/gateway/client.py [hunk 9]
```diff
@@ -95,26 +96,25 @@ def get_route(self, name: str):
...
-                "routes with no `search_filter` defined."
-            )
-        response = self._call_endpoint("GET", MLFLOW_GATEWAY_CRUD_ROUTE_BASE).json()["routes"]
-        return [Route(**resp) for resp in response]
+        request_parameters = None
+        if page_token is not None:
+            request_parameters = {"page_token": page_token}
+        response_json = self._call_endpoint(
+            "GET", MLFLOW_GATEWAY_CRUD_ROUTE_BASE, json_body=json.dumps(request_parameters)
+        ).json()
```
Comment (harupy) [inline mlflow/gateway/client.py:114]: /ai what does `json.dumps(None)` returns in Python?

Comment (dbczumar) [reply mlflow/gateway/client.py:114]: Would rather not pass the null field. Let me know if you strongly disagree

File: mlflow/gateway/constants.py (hunk 7 above)
Comment (dbczumar) [reply mlflow/gateway/constants.py]: Great suggestion! Changed!

File: tests/gateway/test_client.py (hunk 5 above)
Comment (dbczumar) [reply tests/gateway/test_client.py]: Fixed!

File: mlflow/gateway/utils.py (hunk 8 above)
Comment (dbczumar) [reply mlflow/gateway/utils.py]: Great catch - fixed!

**Review by dbczumar — COMMENTED** [review]
@harupy Thank you for the thorough review. I've addressed your comments - can you take another look?

File: mlflow/gateway/client.py (hunk 9 above)
Comment (harupy) [reply mlflow/gateway/client.py:114]: Can we pass `None` to `json_body` if `request_parameters` is None?

**Review by harupy — APPROVED** [review]
LGTM!

File: mlflow/gateway/client.py (hunk 9 above)
Comment (dbczumar) [reply mlflow/gateway/client.py:114]: Ah, yes - fixed

**Pull request merged**